File: py/interview/kvstore_metrics.py
"""
KVStore with metrics

DBricks R1

Make a key value store, but you can mesaure the average QPS over X seconds
"""

import time
from typing import Dict


class KVStore:
    def __init__(self, window_seconds=300):
        self.store: Dict[str, int] = {}
        self.window_seconds = window_seconds
        self.put_buckets = [0 for _ in range(window_seconds)]
        self.get_buckets = [0 for _ in range(window_seconds)]
        self.last_update = int(time.time())

    def put(self, key: str, val: int):
        curr_time_int = int(time.time())
        self._update_buckets(curr_time_int)

        self.store[key] = val

        self.put_buckets[self._get_bucket_idx(curr_time_int)] += 1

    def get(self, key: str) -> int:
        curr_time_int = int(time.time())
        self._update_buckets(curr_time_int)
        self.get_buckets[self._get_bucket_idx(curr_time_int)] += 1

        return self.store[key]

    def put_qps(self) -> float:
        curr_time_int = int(time.time())
        self._update_buckets(curr_time_int)

        return sum(self.put_buckets) / self.window_seconds

    def get_qps(self) -> float:
        curr_time_int = int(time.time())
        self._update_buckets(curr_time_int)

        return sum(self.get_buckets) / self.window_seconds

    def _get_bucket_idx(self, curr_time_int: int) -> int:
        return curr_time_int & self.window_seconds

    def _update_buckets(self, curr_time_int: int):
        if curr_time_int - self.last_update >= self.window_seconds:
            self.put_buckets = [0 for _ in range(self.window_seconds)]
            self.get_buckets = [0 for _ in range(self.window_seconds)]
        else:
            time_diff = curr_time_int - self.last_update

            for i in range(min(time_diff, self.window_seconds)):
                bucket_to_clear = (self.last_update + i + 1) % self.window_seconds
                self.put_buckets[bucket_to_clear] = 0
                self.get_buckets[bucket_to_clear] = 0

        self.last_update = curr_time_int

File: rs/src/interview/kvstore_metrics.rs
use std::{
    cmp::min,
    collections::HashMap,
    hash::Hash,
    time::{SystemTime, UNIX_EPOCH},
};

#[allow(dead_code)]
struct KVStore<K, V> {
    store: HashMap<K, V>,
    put_buckets: Vec<usize>,
    get_buckets: Vec<usize>,
    last_update: u64,
    metric_window: usize,
}

#[allow(dead_code)]
impl<K, V> KVStore<K, V>
where
    K: Eq + Hash,
    V: Clone
{
    pub fn new(metric_window: usize) -> Self {
        Self {
            store: HashMap::new(),
            put_buckets: vec![0; metric_window],
            get_buckets: vec![0; metric_window],
            last_update: SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs(),
            metric_window: metric_window,
        }
    }

    pub fn put(&mut self, key: K, val: V) {
        let now = Self::now();
        self.update_buckets(now);

        let bucket_idx = self.get_bucket_index(now);
        self.put_buckets[bucket_idx] += 1;

        self.store.insert(key, val);
    }

    pub fn get(&mut self, key: K) -> Option<V> {
        let now = Self::now();
        self.update_buckets(now);

        let bucket_idx = self.get_bucket_index(now);
        self.get_buckets[bucket_idx] += 1;

        self.store.get(&key).cloned()
    }

    pub fn put_qps(&mut self) -> f64 {
        let now = Self::now();
        self.update_buckets(now);

        return self.put_buckets.iter().sum::<usize>() as f64 / self.metric_window as f64
    }

    pub fn get_qps(&mut self) -> f64 {
        let now = Self::now();
        self.update_buckets(now);

        return self.get_buckets.iter().sum::<usize>() as f64 / self.metric_window as f64
    }

    fn now() -> u64{
        SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs()
    }

    fn get_bucket_index(&self, now: u64) -> usize {
        return (now % self.metric_window as u64) as usize
    }

    fn update_buckets(&mut self, now: u64) {
        let elapsed = now - self.last_update;

        if elapsed >= self.metric_window as u64 {
            self.put_buckets.fill(0);
            self.get_buckets.fill(0);
        } else {
            let time_diff = now - self.last_update as u64;

            for i in 0..min(time_diff, self.metric_window as u64) {
                let bucket_to_clear = ((self.last_update + i + 1) % self.metric_window as u64) as usize;

                self.put_buckets[bucket_to_clear] = 0;
                self.get_buckets[bucket_to_clear] = 0;
            }
        }

        self.last_update = now;
    }
}
